Remove dependence on `_io` in `collection.py`.

thvoigtmann opened the pull request:
*Update:* please see [comment below](https://github.com/scikit-image/scikit-image/pull/6291#issuecomment-1080509677) for an updated summary that reflects the last commits to this branch.

Adapted `ImageCollection` and `MultiImage` to not rely on the internal
`imread` function of `skimage.io`, but use instead the implementations
provided by `imageio` and `tifffile`.

## Description

This is intended to pick up the discussion around deprecating `skimage.io`, but keeping a thin wrapper around `imageio` as mentioned in [#5439 (comment)](https://github.com/scikit-image/scikit-image/issues/5439#issuecomment-1046269796). As a first step, this PR would remove the dependence of `skimage.io.ImageCollection` and `skimage.io.MultiImage` on the internal implementation of `imread` by `skimage.io`. `ImageCollection` might be nice to keep as it implements functionality that is not obviously covered by other modules (see also [#6276 (PR)](https://github.com/scikit-image/scikit-image/pull/6276)). `MultiImage` probably can be deprecated as its functionality is easy to reproduce with `imageio.mimread`.

For the time being, here I rely on `imageio.imread` to handle most file types, but `tifffile.imread` specifically for the handling of TIFF files (specifically ones with multiple pages); the special handling of TIFF files triggered by any file ending with `.tif` or `.tiff` is inherited from the existing implementation of `ImageCollection`. This might be a point for further improvement.

I am happy to hear comments, @mkcor @stefanv (and if you can point others who have spent more time thinking about this than me).

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

thvoigtmann committed:
Remove dependence on `_io` in `collection.py`.

Adapted `ImageCollection` and `MultiImage` to not rely on the internal
`imread` function of `skimage.io`, but use instead the implementations
provided by `imageio` and `tifffile`.

thvoigtmann committed:
Fixed line-too-long issue.

stefanv commented:
@thvoigtmann I'm not sure I follow the logic of this PR.  ImageCollections load a variety of different file types together, so you'd have to provide a specialized load_func if you want to handle tiff files in a special way.

thvoigtmann commented:
@stefanv I totally agree that a custom `load_func` should be the way to treat some file (formats) specially. It just happens that the existing implementation of `ImageCollection` handles TIFF files in a special way, and so far I was not touching that piece of logic.

thvoigtmann committed:
Removed explicit dependence on tifffile.

Removed the separate test for files ending in `.tif` or `.tiff` in
`ImageCollection`. This does not change the behavior in how TIFF files
are treated.

thvoigtmann committed:
Update to use imageio.v3 API.

This fixes the loading of multi-frame GIF images.

thvoigtmann committed:
Merge branch 'main' into io-collection-mod

thvoigtmann committed:
Fixed MultiImage with new imageio interface.

To recover the original behavior of `MultiImage` combined with the
internal `_io` library, implemented a small wrapper around
`imageio.v2.mimread`.

thvoigtmann committed:
Fixed MultiImage behavior to make tests pass.

thvoigtmann committed:
Fixed `MultiImage` tests.

thvoigtmann commented:
@stefanv if you could have a look again, I'd be happy.

With the latest commit from this branch, both `ImageCollection` and `MultiImage` got rid of the internal `_io` module, and also no longer involve any special treatment for specific file formats. I find this now much cleaner and hopefully more consistent.

The loading of images is completely relayed to `imageio.v3`, unless a custom ``load_func`` is specificed.

There is still WIP to adapt the initialization of `ImageCollection` that currently is rather slow when working with a large collection of files.

thvoigtmann committed:
Removed explicit dependence on PIL.

Rewrote the `ImageCollection` initialization code to use the `imageio`
interface instead of `PIL`.

thvoigtmann committed:
Fixed failing tests.

thvoigtmann committed:
Fixed whitespace issue.

thvoigtmann committed:
Added backwards-compatibility with imageio v2.

With this commit, code is added to provide silent fall-back to using
the old-style imageio API (pre-2.10.0) in case loading of imageio.v3
fails.

thvoigtmann commented:
## Updated summary

Adapted `ImageCollection` and `MultiImage` to rely on the `imageio` module for the handling of IO, dropping the need for an internal `skimage.io._io`.

## Updated Description

This is intended to pick up the discussion around deprecating `skimage.io`, but keeping a thin wrapper around `imageio` as mentioned in [#5439 (comment)](https://github.com/scikit-image/scikit-image/issues/5439#issuecomment-1046269796). `ImageCollection` might be nice to keep as it implements functionality that is not obviously covered by other modules (see also [#6276 (PR)](https://github.com/scikit-image/scikit-image/pull/6276)). `MultiImage` probably can be deprecated as its functionality is easy to reproduce with `ImageCollection` and a custom `load_func`.

With this PR:
- `ImageCollection` and `MultiImage` work independently of the internal IO functions that are still kept in `skimage.io`, so that the latter can be deprecated.
- Image IO is relayed to `imageio.v3` if available, with a fallback to old-style `imageio`.
- Side effect: The handling of multi-frame images is more consistent than before. The previous (`skimage.io`-internal) implementation had specific handling for files with names ending in `.tif` or `.tiff` and did not consistently read multiple frames from PNG images in `MultiImage`. The implementation here relies on `imageio` to handle the detection of multiple frames on the file-type specific plugin level. It still detects multi-frame TIFF files as before, and also reads multiple frames from PNG files.
- The previous implementation also depended on `PIL` for the detection of the number of frames per image file. The code here replaces that by the corresponding `imageio` methods. This should be more consistent (and having fewer explicit requirements), since I could imagine situations where `PIL` might detect a different number of frames than the corresponding `imageio` plugin.
- Side effect (change in behavior): `MultiImage` returns arrays of different shape now. See below.

The intention of `ImageCollection` seems clear to me: map a list of image filenames to a list of image data, allowing random access to the images. If a file contains multiple frames, each frame is mapped to a separate entry in the list of image data. So, `len(ImageCollection(gif_or_tiff_file))` is equal to the number of frames in the corresponding file.

`MultiImage` in the previous implementation returned `1` as the length if instantiated with a single TIFF file containing `N` frames of width and height `W x H`, and stored all the data in a single image-data entry of shape `(N, W, H)`. Except if the file contains only one frame, in which case the image data was stored as an array of shape `(W, H)`. I find this debatable. It appears that `MultiImage` was created with just TIFF files in mind, and meant to be invoked with a single multi-frame file only.

The implementation here makes `MultiImage` in my opinion more consistent: it will always store the frames of each file in an array of shape `(N, W, H)`, even if `N==1`.

thvoigtmann committed:
Merge remote-tracking branch 'upstream/main' into io-collection-mod

FirefoxMetzger commented:
> The implementation here makes MultiImage in my opinion more consistent: it will always store the frames of each file in an array of shape (N, W, H), even if N==1.

How do you decide if a file contains a single image or multiple images, i.e., when to set N=1 and when to choose N from the file? At first glance, this seems to work for simple images (JPG, PNG, GIF) and for video, but how would you handle volumetric images, channel-first TIFFs (e.g. staining pictures), time-series, and similar types of image data?

What's more, how will this be handled in a multi-file/folder scenario, where you may encounter any combination of the above?

thvoigtmann commented:
Good point @FirefoxMetzger - the more I think about it, the less obvious it becomes to me, whether to keep `MultiImage` around at all. It seems to add little functionality to existing `imageio` functions.

> How do you decide if a file contains a single image or multiple images, i.e., when to set N=1 and when to choose N from the file? At first glance, this seems to work for simple images (JPG, PNG, GIF) and for video, but how would you handle volumetric images, channel-first TIFFs (e.g. staining pictures), time-series, and similar types of image data?

I wanted to leave that decision to the underlying `imiter` from `imageio.v3` by default, unless the user knows better and specifies their own `load_func`. So, by default, `N = len([*imiter(file)])`.

> What's more, how will this be handled in a multi-file/folder scenario, where you may encounter any combination of the above?

I'm not sure I understand that point. Each entry in a `MultiImage` can have completely different shape.

stefanv commented:
@FirefoxMetzger Thanks again for your work on this. All above looks good to me, except:

> If a file contains multiple frames, each frame is mapped to a separate entry in the list of image data.

I'd expect `len(ImageCollection(fn_list)) == len(fn_list)`, and given @thvoigtmann's comment it's not clear we can honor that commitment with shape-varying multi-images.  In that case, it may be best to error out.  For multi-image types where all images have the same shape, it'd be fine to return a single array per multi-image.

thvoigtmann commented:
> I'd expect `len(ImageCollection(fn_list)) == len(fn_list)`, and given @thvoigtmann's comment it's not clear we can honor that commitment with shape-varying multi-images. In that case, it may be best to error out. For multi-image types where all images have the same shape, it'd be fine to return a single array per multi-image.

I suggest to add a flag to let the user decide. This will have another beneficial side effect that I find important.

To explain:

I found it reasonable to demand that `ImageCollection` should focus on functionality that isn't straightforwardly provided by `imageio`. To me, this mostly leaves the use case as a wrapper around a large set of images to which you want to have random access, but that you do not want to store all in memory at once.

For that reason, I would expect `ImageCollection` to index all pages/frames in multi-image types separately (only ever storing one in memory at the expense of performance). This however leads to `len(ImageCollection(fn_list)) >= len(fn_list)`.

If the user wants a single array per multi-image, it seems that either `imageio` or `ImageCollection` with a custom `load_func` can do the job. In fact, the current `MultiImage` is exactly that. I should check, but I tink that `assert len(MultiImage(fn_list)) == len(fn_list))` is `True`, except for shell globbing. (The underlying `ImageCollection` performs shell expansion on `fn_list`, and I didn't touch that part of the code.)

None of these care about varying shapes, and I don't see why they should. Am I overlooking something @stefanv? If you'd need that for performance reasons, getting a single file's data with one of the `imageio` functions into a numpy array will be the way to go.

There is one drawback: `ImageCollection` needs to figure out how many pages/frames there are, so it needs to go through all files in `fn_list` upon instantiation. One of my use cases involves some ten thousand files on a slow storage medium, and that's painful. I was thinking about implementing lazy instantiation for that case.

So, one could add a flag to let the user decide: either "index per frame" (like now: `ic[i]` always refers to a single page/frame of a file, but instantiation of `ic = ImageCollection(fn_list)` is costly), or "index per file" (`ic[i]` refers to whatever the `imageio` plugin decides, and instantiation is lazy because we don't need to access the files outside `__getitem__`).

I'd be happy to code that if you want.

thvoigtmann committed:
Added index_per_frame flag.

This allows to control the behavior of `ImageCollection` regarding
multi-frame image files such as TIFF or GIF: let the user decide
whether an `ImageCollection` entry should correspond to a single
frame in a file (default), or to the whole file at once.

FirefoxMetzger commented:
> I'm not sure I understand that point. Each entry in a MultiImage can have completely different shape.

@thvoigtmann I agree. I was asking, because you said that the user would always get a `(N, W, H)` shaped image and I couldn't quite follow how this would be realized.

> For multi-image types where all images have the same shape, it'd be fine to return a single array per multi-image.

@stefanv Hmm, I'm not sure I got this part right. What I currently understand is if I have a TIFF that represents a voxel image then `ImageCollection` should return that as a `(stack, height, width)` image; however, if I have a TIFF that is simply a stack of pages with same shape (no particular flavor, just plain old TIFF), should it stack those as well or should it raise an exception?

> For that reason, I would expect ImageCollection to index all pages/frames in multi-image types separately (only ever storing one in memory at the expense of performance).  [...] If the user wants a single array per multi-image, it seems that either imageio or ImageCollection with a custom load_func can do the job.

I'm not sure that this is the intuitive path for most users. Two examples: 

(1) TIFF is one of the more "misused" formats out there, meaning that there is the official spec, and then there is what people actually do with it. For example, ImageJ has a thing they call hyperstacks, which are essentially TIFF files with metadata that is saved in the comment text and this metadata tells you how to reassemble the image/hyperstack from the pages that are stored in the TIFF. Without this metadata the individual planes may not make much sense, because they could be different z-levels, different color planes, or different time slices in a series and you wouldn't know. Many people I've talked with think of hyperstacks as one block and serving them individual slices would make them think the reader is somehow broken. The same is true for other flavors of TIFF; most of the time individual pages is not what the user is thinking about.

(2) A colleague of mine does deep learning on video data and for that he has sliced a video into small 5s clips. For him, these clips are a single unit. You can identify individual frames as units or images and see the video as a sequence of images, but for him that would not make sense and he would prefer to read the video as a single ndimage of shape `(time, height, width, channel)`. This is true not just for him, but for many people in the video learning community, where videos in benchmark datasets have a set start and end interval for each video. Admittedly this is video and not images, but from the point of ndimages they are quite related.

My point here is: For multi-images, the unit/shape a user expects to get back is usually not the unit the container uses to store pixel data (pages for TIFF), but rather something more high-level. What that something is is very application-specific and it's hard/impossible to say in general, but it's usually very clear and intuitive once you have the context. You can choose to serve individual pages, slices, frames, etc. by default, but this choice will probably be wrong more often than it is right, which makes it a not-so-ideal default. I think the safer thing would be to raise by default when a multi-image format is encountered and to provide an easy way to customize this behavior and instead serve pages/slices if desired.

thvoigtmann commented:
@FirefoxMetzger I understand your point. The video example however doesn't convince me, because I think in this case you'd provide a custom `load_func` that slices the video into the segments that you desire. See the discussion around [#6276 (PR)](https://github.com/scikit-image/scikit-image/pull/6276). If you have a nice small-ish example video, one could provide a corresponding example in the docstring.

Regarding your proposal to raise when a multi-image format is encountered: what would be a good way of detecting that? In `imageio.v3` I can imagine calling `imiter` and raising whenever that returns more than one element. In `imageio.v2` I am not so sure.

FirefoxMetzger commented:
> The video example however doesn't convince me, because I think in this case you'd provide a custom load_func that slices the video into the segments that you desire. 

Oh, sorry, I was talking about loading the entire file in one go which contains 5s worth of video. Either way, I agree that a custom `load_func` is likely the most clear path forward.

> Regarding your proposal to raise when a multi-image format is encountered: what would be a good way of detecting that?

I'd go with [`iio.improps`](https://imageio.readthedocs.io/en/stable/_autosummary/imageio.v3.improps.html#imageio.v3.improps) which returns a dataclass of type [`ImageProperties`](https://imageio.readthedocs.io/en/stable/_autosummary/imageio.core.v3_plugin_api.ImageProperties.html?highlight=ImageProperties#imageio.core.v3_plugin_api.ImageProperties). This is part of the v3 API and will try to avoid reading pixel data where possible (limited to v3 plugins; v2 plugins will read pixels either way). In code, this could look something like

```python
props = iio.improps(image, index=...)
if props.is_batch and props.shape[0] > 1:
    raise ValueError("The file contains more than one ndimage.")
```

Alternatively, you could do something less verbose which simply checks if the default `imread` would yield more than one ndimage:

```python
if iio.improps(image).is_batch:
    raise ValueError("`iio.imread(image)` would load more than one ndimage.")
```

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
